### PR TITLE
Meta protocol updates

### PR DESCRIPTION
**Pull request by dormando**
Commits fleshing out the TODO's from #546 

- fixes a refcount leak (oops!)
- some missing tests/cleanup
- refactors some core code
- adds the mode switch to metaset: set, add, cas, replace, append, prepend are now all just the `ms` command.
- adds meta arithmetic (`ma`) command.

### CODE REVIEW

**Comment by dormando** [comment]
@shivnagarajan you might be interested in this

File: memcached.c [hunk 1]
```diff
@@ -4136,18 +4209,62 @@ static int _meta_flag_preparse(token_t *tokens, const size_t ntokens, struct _me
...
+                }
+                break;
+            case 'J': // marithmetic initial value
+                if (!safe_strtoull(tokens[i].value+1, &of->initial)) {
+                    *errstr = "CLIENT_ERROR invalid numeric initial value";
+                    of->has_error = 1;
+                }
+                break;
+            case 'D': // marithmetic delta value
+                if (!safe_strtoull(tokens[i].value+1, &of->delta)) {
```
Comment (shivnagarajan) [inline memcached.c]: So the `M` mode option for `ma` accepts `I, `D`, `+` or `-`,  and assumes implicitly that the `D` option will not accept negative values (which `safe_strtoull` handles and appropriately fails).

But the protocol description does say that the argument for `-D` is an `unsigned int`, but `safestrtoull` will happily accept a string value > an `unsigned int`. Maybe this is not that important, but we could easily adapt the protocol documentation to just say `will accept an argument that is a positive unsigned long long` maybe? 🤷‍♂

File: memcached.c [hunk 2]
```diff
@@ -4563,29 +4642,51 @@ static void process_mset_command(conn *c, token_t *tokens, const size_t ntokens)
...
+    // APPEND/PREPEND allow a simplified CAS check.
+    if (of.has_cas && (comm == NREAD_SET || comm == NREAD_REPLACE)) {
+        comm = NREAD_CAS;
+    }
 
-    // TODO: can we treat vlen as unsigned? :(
-    if (vlen < 0 || vlen > (INT_MAX - 2)) {
-        // TODO: specific error.
+    // We attempt to process as much as we can in hopes of getting a valid and
+    // adjusted vlen, or else the data swallow after error will be for 0b.
```
Comment (shivnagarajan) [inline memcached.c]: typo `swallow` -> `swallowed`

File: doc/protocol.txt [hunk 3]
```diff
@@ -727,12 +728,23 @@ See description under 'Meta Get'
...
 
 Errors are always returned.
 
+- M(token): mode switch. Takes a single character for the mode.
+
+E: "add" command. LRU bump and return NS if item exists. Else
+add.
+A: "append" command. If item exists, append the new value to its data.
+P: "prepend" command. If item exists, prepend the new value to its data.
+S: "set" command. The default mode, added for completeness.
```
[the remark this answers is not in the document]
Comment (dormando) [reply doc/protocol.txt]: Fixed, thanks!

File: doc/protocol.txt [hunk 4]
```diff
@@ -795,6 +807,88 @@ When marking an item as stale with 'I', the 'T' flag can be used to update the
...
+- T(token): update TTL
+
+On success, sets the remaining TTL to the supplied value.
+
+-M(token): mode switch. Takes a single character for the mode.
+
+I: Increment mode (default)
++: Alias for increment
+D: Decrement mode
+-: Alias for decrement
```
[the remark this answers is not in the document]
Comment (dormando) [reply doc/protocol.txt]: It does have one, I missed the doc. In meta commands "noreply" is a separate q flag. There's even a test for it in metaget.t :) Updated locally now.

File: memcached.c (hunk 1 above)
Comment (dormando) [reply memcached.c]: Nice catch! This was supposed to match the incr/decr documentation as it handles arguments the same way. I've updated a few words to explicitly say 64-bit unsigned integer.

File: memcached.c (hunk 2 above)
Comment (dormando) [reply memcached.c]: fixed

**Comment by dormando** [comment]
Found another bug this morning + doc fixes from review comments. Think I'm done with this and ready to merge into `next`.

**Comment by dormando** [comment]
Thanks :) I'm excited to get this out there.

**Pull request merged**